### PR TITLE
Use KMS in Sam

### PR DESCRIPTION
**Pull request by marctalbott**
Ticket: [GAWB-4037](https://broadinstitute.atlassian.net/browse/GAWB-4037)
- This creates a KMS key and keyring and grants access to that key to the All_Users group

---

**PR checklist**

- [x] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala [hunk 1]
```diff
@@ -134,6 +128,25 @@ class GoogleExtensions(
...
+      _ <- googleKms.createKeyRing(GoogleProject(googleServicesConfig.googleKms.project),
+        Location(googleServicesConfig.googleKms.location),
+        KeyRingId(googleServicesConfig.googleKms.keyRingId)) handleErrorWith { case _: AlreadyExistsException => IO.unit }
+
+      _ <- googleKms.createKey(GoogleProject(googleServicesConfig.googleKms.project),
+        Location(googleServicesConfig.googleKms.location),
+        KeyRingId(googleServicesConfig.googleKms.keyRingId),
+        KeyId(googleServicesConfig.googleKms.keyId),
+        Option(Timestamp.newBuilder().setSeconds(System.currentTimeMillis() / 1000 + 180.days.toSeconds).build()),
+        Option(Duration.newBuilder().setSeconds(180.days.toSeconds).build())
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala]: make 180 days configurable

**Review by dvoet — CHANGES_REQUESTED** [review]
1 config change then 👍

File: src/test/resources/reference.conf [hunk 2]
```diff
@@ -55,6 +55,13 @@ googleServices {
       workerCount = 1
     }
   }
+
+  kms {
+    project = "broad-dsde-dev"
+    location = "global"
+    keyRingId = "mtalbott"
+    keyId = "mtalbott"
```
Comment (Qi77Qi) [inline src/test/resources/reference.conf]: shall we change `keyRingId` and `keyId` to something else that's not specific to you

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/config/KmsConfig.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  * @param location location for the key
+  * @param keyRingId name of the key ring that contains the key
+  * @param keyId name of the key
+  */
+
+final case class GoogleKmsConfig(
+                                  project: String,
+                                  location: String,
+                                  keyRingId: String,
+                                  keyId: String
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/config/KmsConfig.scala]: seems like there're custom types for `project`, `location`, `keyRingId` etc, why not use those? that way, you won't need to wrap them [here](https://github.com/broadinstitute/sam/pull/275/files#diff-ab230e3b899f0b14c89b226c3c761c0eR138). It's also more type safe

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/config/KmsConfig.scala [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  * @param keyRingId name of the key ring that contains the key
+  * @param keyId name of the key
+  */
+
+final case class GoogleKmsConfig(
+                                  project: GoogleProject,
+                                  location: Location,
+                                  keyRingId: KeyRingId,
+                                  keyId: KeyId,
+                                  rotationPeriod: Int
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/config/KmsConfig.scala]: what is `Int`? if this is duration, we should probably use `FiniteDuration`

**Review by Qi77Qi — APPROVED** [review]
one minor comment

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/config/KmsConfig.scala (hunk 4 above)
Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/config/KmsConfig.scala]: It's the number of days. I call `.days.toSeconds` on it. I suppose we could use a Duration, but the number of days as an int was what I thought of first

Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/config/KmsConfig.scala]: so lots config libraries support reading `FiniteDuration` directly. x days probably can be `x day` in config directly...see https://github.com/iheartradio/ficus#examples, there's an example for `minutes`, but I think it should definitely have `day` too

```
  // something such as "15 minutes" can be converted to a FiniteDuration
  val retryInterval: FiniteDuration = config.as[FiniteDuration]("retryInterval")
```

The benefit of this is we don't need to guess what unit this `Int` is anymore. You can also just do `toSeconds` at call site

Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/config/KmsConfig.scala]: did a bit more digging, I see https://github.com/iheartradio/ficus/blob/a6db5d65f3c3237a27aee592661678962a2aca09/src/test/scala/net/ceedubs/ficus/readers/DurationReadersSpec.scala#L34, I think you should be able to do `180 days` in config file and read it as a `FiniteDuration`

**Review by dvoet — APPROVED** [review]

**Pull request merged**